cmd/metal-api/internal/scaler/poolscaler.go

```diff
@@ -0,0 +1,178 @@
+package scaler
+
+import (
+	"math"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
+	"go.uber.org/zap"
+)
+
+type (
+	PoolScaler struct {
+		log     *zap.SugaredLogger
+		manager MachineManager
+	}
+	PoolScalerConfig struct {
+		Log     *zap.SugaredLogger
+		Manager MachineManager
+	}
+	MachineManager interface {
+		AllMachines() (metal.Machines, error)
+		WaitingMachines() (metal.Machines, error)
+		ShutdownMachines() (metal.Machines, error)
+
+		Shutdown(m *metal.Machine) error
+		PowerOn(m *metal.Machine) error
+	}
+)
+
+func NewPoolScaler(log *zap.SugaredLogger, manager MachineManager) *PoolScaler {
+	return &PoolScaler{
+		log:     log,
+		manager: manager,
+	}
+}
+
+// AdjustNumberOfWaitingMachines compares the number of waiting machines to the required pool size of the partition and shuts down or powers on machines accordingly
+func (p *PoolScaler) AdjustNumberOfWaitingMachines(partition *metal.Partition) error {
+	waitingMachines, err := p.manager.WaitingMachines()
+	if err != nil {
+		return err
+	}
+
+	poolSizeExcess := 0
+	poolSizeExcess = p.calculatePoolSizeExcess(len(waitingMachines), partition.WaitingPoolMinSize, partition.WaitingPoolMaxSize)
+
+	if poolSizeExcess == 0 {
+		return nil
+	}
+
+	if poolSizeExcess > 0 {
+		p.log.Infow("shut down spare machines", "number", poolSizeExcess, "partition", p)
+		err := p.shutdownMachines(waitingMachines, poolSizeExcess)
+		if err != nil {
+			return err
+		}
+	} else {
+		shutdownMachines, err := p.manager.ShutdownMachines()
+		if err != nil {
+			return err
+		}
+
+		poolSizeExcess = int(math.Abs(float64(poolSizeExcess)))
```

Comment:
Is taking the absolute value really required? Isn't the value always >= 0?

Reply:
The function `calculatePoolSizeExcess` returns `actual` number of waiting machines minus `average` of max and min required waiting machines. It will be the positive value of machines, that need to be shut down, if pool size is exceeded and the negative value of machines, that need to be powered on, if more machines are needed.

Reply:
Okay, understood. Thanks.